File: SupermarketApi/Entities/OrderAggregate/Order.cs
#nullable disable
namespace SupermarketApi.Entities.OrderAggregate
{
    using System;
    using System.Collections.Generic;
    using static OrderStatus;

    public class Order : BaseEntity
    {
        public Order()
        {
        }

        public Order(
            IReadOnlyList<OrderItem> orderItems,
            string buyerEmail,
            Address shipToAddress,
            DeliveryMethod deliveryMethod,
            decimal subtotal)
        {
            this.OrderItems = orderItems;
            this.BuyerEmail = buyerEmail;
            this.DeliveryMethod = deliveryMethod;
            this.ShipToAddress = shipToAddress;
            this.Subtotal = subtotal;
        }

        public string BuyerEmail { get; set; }

        public DeliveryMethod DeliveryMethod { get; set; }

        public DateTimeOffset OrderDate { get; set; } = DateTimeOffset.Now;

        public IReadOnlyList<OrderItem> OrderItems { get; set; }

        public OrderStatus Status { get; set; } = Pending;

        public string PaymentIntentId { get; set; }

        public Address ShipToAddress { get; set; }

        public decimal Subtotal { get; set; }

        public decimal GetTotal()
        {
            return this.Subtotal + this.DeliveryMethod.Price;
        }
    }
}

File: Client/src/app/shared/models/order.ts
import { IAddress } from './address';

export interface IOrderToCreate {
  basketId: string;
  deliveryMethodId: number;
  shipToAddress: IAddress;
}

export interface IOrder {
  id: number;
  buyerEmail: string;
  deliveryMethod: string;
  orderDate: string;
  orderItems: IOrderItem[];
  status: string;
  shippingPrice: number;
  shipToAddress: IAddress;
  subtotal: number;
  total: number;
}

export interface IOrderItem {
  pictureUrl: string;
  price: number;
  productId: number;
  productName: string;
  quantity: number;
}
